src/device/socket/vsock.rs

```diff
@@ -101,24 +139,185 @@ impl<'a, H: Hal, T: Transport> VirtIOSocket<'a, H, T> {
             op: VirtioVsockOp::Request.into(),
             ..Default::default()
         };
-        self.tx
-            .add_notify_wait_pop(&[header.as_bytes(), &[]], &mut [], &mut self.transport)?;
+        self.send_packet_to_tx_queue(&header, &[])?;
+        let received_header = self.pop_header_only_packet_from_rx_queue()?;
+        match received_header.op()? {
+            VirtioVsockOp::Response => {
+                let dst = VsockAddr {
+                    cid: dst_cid,
+                    port: dst_port,
+                };
+                self.connection_info = Some(ConnectionInfo {
+                    dst,
+                    src_port,
+                    ..Default::default()
+                });
+                Ok(())
+            }
+            VirtioVsockOp::Rst => Err(SocketError::ConnectionFailed.into()),
+            VirtioVsockOp::Shutdown => Err(SocketError::PeerSocketShutdown.into()),
+            _ => Err(SocketError::InvalidOperation.into()),
+        }
+    }
+
+    /// Requests the credit and updates the credit in the current connection info.
+    pub fn request_credit(&mut self) -> Result {
+        let connection_info = self.connection_info()?;
+        let header = VirtioVsockHdr {
+            src_cid: self.guest_cid.into(),
+            dst_cid: connection_info.dst.cid.into(),
+            src_port: connection_info.src_port.into(),
+            dst_port: connection_info.dst.port.into(),
+            op: VirtioVsockOp::CreditRequest.into(),
+            ..Default::default()
+        };
+        self.send_packet_to_tx_queue(&header, &[])?;
+        self.wait_one_in_rx_queue();
+        let received_header = self.pop_header_only_packet_from_rx_queue()?;
```

Comment:
Is it guaranteed that the next packet we receive after sending a credit request will be a credit update, or could it be interleaved with other packets?

Reply:
It's possible actually. There're still some virtqueue controls not yet implemented, like I should process all the rx queue before sending out the request, and process the rx queue again when a packet arrives. That needs to be done for all these functions.

I would like to make the communication work first in this PR, then address these scenarios in a follow-up change. I added a TODO.